src/monodroid/jni/embedded-assemblies-zip.cc

```diff
@@ -2,79 +2,131 @@
 #include <cerrno>
 #include <cctype>
 #include <vector>
-#include <type_traits>
 #include <libgen.h>
 
 #include <mono/metadata/assembly.h>
 
 #include "embedded-assemblies.hh"
 #include "globals.hh"
 #include "xamarin-app.hh"
+#include "xxhash.hh"
 
 using namespace xamarin::android::internal;
 
 using read_count_type = size_t;
 
-force_inline bool
-EmbeddedAssemblies::is_debug_file (dynamic_local_string<SENSIBLE_PATH_MAX> const& name) noexcept
-{
-	return utils.ends_with (name, ".pdb");
-}
-
 force_inline bool
 EmbeddedAssemblies::zip_load_entry_common (size_t entry_index, std::vector<uint8_t> const& buf, dynamic_local_string<SENSIBLE_PATH_MAX> &entry_name, ZipEntryLoadState &state) noexcept
 {
 	entry_name.clear ();
 
 	bool result = zip_read_entry_info (buf, entry_name, state);
 
-#ifdef DEBUG
-	log_info (LOG_ASSEMBLY, "%s entry: %s", state.apk_name, entry_name.get () == nullptr ? "unknown" : entry_name.get ());
-#endif
+	log_debug (LOG_ASSEMBLY, "%s entry: %s", state.file_name, entry_name.get () == nullptr ? "unknown" : entry_name.get ());
 	if (!result || entry_name.empty ()) {
-		log_fatal (LOG_ASSEMBLY, "Failed to read Central Directory info for entry %u in APK file %s", entry_index, state.apk_name);
+		log_fatal (LOG_ASSEMBLY, "Failed to read Central Directory info for entry %u in APK file %s", entry_index, state.file_name);
 		Helpers::abort_application ();
 	}
 
-	if (!zip_adjust_data_offset (state.apk_fd, state)) {
-		log_fatal (LOG_ASSEMBLY, "Failed to adjust data start offset for entry %u in APK file %s", entry_index, state.apk_name);
+	if (!zip_adjust_data_offset (state.file_fd, state)) {
+		log_fatal (LOG_ASSEMBLY, "Failed to adjust data start offset for entry %u in APK file %s", entry_index, state.file_name);
 		Helpers::abort_application ();
 	}
-#ifdef DEBUG
-	log_info (LOG_ASSEMBLY, "    ZIP: local header offset: %u; data offset: %u; file size: %u", state.local_header_offset, state.data_offset, state.file_size);
-#endif
+
+	log_debug (LOG_ASSEMBLY, "    ZIP: local header offset: %u; data offset: %u; file size: %u", state.local_header_offset, state.data_offset, state.file_size);
 	if (state.compression_method != 0) {
 		return false;
 	}
 
-	if (entry_name.get ()[0] != state.prefix[0] || strncmp (state.prefix, entry_name.get (), state.prefix_len) != 0) {
-		return false;
+	if (entry_name.get ()[0] != state.prefix[0] || memcmp (state.prefix, entry_name.get (), state.prefix_len) != 0) {
+		if (state.prefix == apk_lib_prefix.data ()) {
```

Comment:
Isn't this just a pointer comparison?

I'm going to have to track what `state.prefix` is, aren't I…

Reply:
They can be the same pointer, [see here](https://github.com/xamarin/xamarin-android/pull/8478/files#diff-a906440c8e0b3839f8bca3c875c6a339c66182d57f344b7db0c66f456e1f9c2aR272).

If they're the same location, we short circuit the checking and avoid some comparisons in the next `if` statement.